virtualization_view.m

```diff
@@ -165,11 +165,30 @@ - (instancetype)init
 
 @end
 
+API_AVAILABLE(macos(12.0))
+@interface VMStateObserver : NSObject
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context;
+@end
+
+@implementation VMStateObserver
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context;
+{
+    if ([keyPath isEqualToString:@"state"]) {
+        int newState = (int)[change[NSKeyValueChangeNewKey] integerValue];
+        if (newState == VZVirtualMachineStateStopped || newState == VZVirtualMachineStateError) {
+            [NSApp performSelectorOnMainThread:@selector(terminate:) withObject:context waitUntilDone:NO];
```

Comment:
I'm not sure why is not enough code lines at 228 and 234?

Reply:
Ah, maybe We don't need state for error...?

Reply:
I think this API is better than observing
https://developer.apple.com/documentation/virtualization/vzvirtualmachinedelegate/3656731-virtualmachine?language=objc

Reply:
This API is already used in Code-Hex/vz https://github.com/Code-Hex/vz/blob/4f8ae6d82e477251a3241bb9696b83959947c5fb/virtualization_view.m#L207-L212
but the log is never printed/the method is never called. The VM did not stop because of an error, maybe it is normal this is not called.

Reply:
Could you try this code with Xcode?
NSLog shows only on the Xcode console

Reply:
So... current code is racing with this API and new observing code that you added.


I think better fix:

1. Remove code in didStopWithError
2. Add code to release observer when dealloc view


And I'm not sure call remove observer in observe handler. (I think this is anti pattern) So it's good calling together with method to release observer

Reply:
Thanks for the suggestions, I'll take a look!

Reply:
> I'm not sure why is not enough code lines at 228 and 234?

(started typing this before your last round of comments, posting it as it adds a bit of explanations)

If we call `Stop` from the host:
- the *guest* did not stop the virtual machine, so `guestDidStopVirtualMachine` is not called
- there was *no error*, so ` didStopWithError` is not called